### ISSUE TITLE
k8s: How to mount manually created cephFS?

### ISSUE THREAD

**Issue by sajal** [rook/rook#567]
So far I have been using rook/rbd using StorageClass, it works fine after cloning the `rook-rbd-user` secret into the namespace.

Now I am looking into cephfs because I need to share a writable filesystem among pods (not really concerned about IO performance much). Specifically I want to run kube-registry with high availability which is achieved by running multiple replicas using same storage volume.

I created a shared filesystem as per the [docs](https://github.com/rook/rook/blob/master/Documentation/client.md#shared-file-system)

A single `mds` pod comes up once I crate the filesystem.

```
root@rook-tools:/# rook filesystem ls                                                                                                                                                               
NAME         METADATA POOL         DATA POOLS
registryFS   registryFS-metadata   registryFS-data
root@rook-tools:/# 
```

My question is how to express this as a volume in the pod container spec?

So far I've got

```
      volumes:
      - name: image-store
        cephfs:
          monitors:
            - mon.rook.svc.cluster.local:6790
          path: /registryFS
          secretRef:
            name: rook-rbd-user
```

PS: I take care of making `mon.rook.svc.cluster.local` resolvable.

And pod events

```
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----			-------------	--------	------		-------
  15s		15s		1	default-scheduler			Normal		Scheduled	Successfully assigned kube-registry-v0-dsh5c to 10.130.48.207
  15s		7s		5	kubelet, 10.130.48.207			Warning		FailedMount	MountVolume.SetUp failed for volume "kubernetes.io/cephfs/8ef6ed50-1f58-11e7-8433-1a1fff709857-image-store" (spec.Name: "image-store") pod "8ef6ed50-1f58-11e7-8433-1a1fff709857" (UID: "8ef6ed50-1f58-11e7-8433-1a1fff709857") with: CephFS: mount failed: mount failed: exit status 32
Mounting command: mount
Mounting arguments: mon.rook.svc.cluster.local:6790:/registryFS /var/lib/kubelet/pods/8ef6ed50-1f58-11e7-8433-1a1fff709857/volumes/kubernetes.io~cephfs/image-store ceph [name=admin,secret=AQSNIPPED==]
Output: mount: permission denied
```

**Comment by jbw976** [comment]
@sajal, i think the issue here is that the RBD cephx user that rook creates does not get any permissions for MDS so it is not allowed to mount the ceph filesystem.

You could work around this by using the admin user instead (user name is just `admin` and the secret name is `rook-admin`).  We're also working on seeing if there's a more complete example for you...

**Comment by sajal** [comment]
thanks for the tip, I got it to work. 

For reference this is steps I took.

1. Create filesystem

`rook filesystem create --name registryFS`
2. Adjust replication

```
ceph osd pool set registryFS-data size 2
ceph osd pool set registryFS-metadata size 2
```
3. Copy admin secret to the desired namespace

`kubectl get secret rook-admin -n rook -o json | jq '.metadata.namespace = "kube-system"'  | kubectl apply -f -`

4. Pod volume spec - adapted from [ceph kubernetes docs](https://raw.githubusercontent.com/ceph/ceph-docker/master/examples/kubernetes/ceph-cephfs-test.yaml)

```
      volumes:
      - name: image-store
        cephfs:
          monitors:
          - mon.rook.svc.cluster.local:6790
          user: admin
          secretRef:
            name: rook-admin
```

I do not know how to tell rook/ceph about which filesystem to use, but since `registryFS` is the only one created I guess thats used by default. I verified by checking `rook filesystem mount` and my files are in it.

I now have a multi-replica kube-registry ...

**Comment by jbw976** [comment]
@sajal this is actually really cool! this is such an interesting example of using rook/ceph storage in the cluster as a backing store for a useful cluster service...i still need to understand better some of the pod/storage lifecycle implications, but i did want to give kudos for how cool this is :)

our docs only have a block storage example, so a good filesystem example would be a much appreciated addition.  we should think about how best to incorporate your learnings into official repo documentation, it would be very useful.  i'll think more about that, but if you submitted a PR with what you know so far, i'm sure we could help it get to the finish line.  thanks again @sajal!

**Comment by sajal** [comment]
Will try add documentation.

In an ideal world, I could create TPR for filesystem and object storage, and consume them without DNS hacks.

**Comment by kokhang** [comment]
@sajal that is part of the roadmap. I dont think there is an issue related to FS and object TPR but feel free to propose one and mentioned what you think it should look like :)

**Comment by travisn** [comment]
Fixed by #578
